core/src/main/java/org/apache/iceberg/PartitionStatsUtil.java

```diff
@@ -0,0 +1,218 @@
+   * @param manifest the manifest file containing metadata about the records.
+   * @param recordSchema the schema defining the structure of the records.
+   * @return a CloseableIterable of partition stats records as defined by the manifest file and
+   *     record schema.
+   */
+  public static CloseableIterable<Record> fromManifest(
+      Table table, ManifestFile manifest, Schema recordSchema) {
+    return CloseableIterable.transform(
+        ManifestFiles.open(manifest, table.io(), table.specs())
+            .select(BaseScan.scanColumns(manifest.content()))
+            .liveEntries(),
+        entry -> fromManifestEntry(entry, table, recordSchema));
+  }
+
+  /**
+   * Appends statistics from one Record to another.
+   *
+   * @param toRecord the Record to which statistics will be appended.
+   * @param fromRecord the Record from which statistics will be sourced.
+   */
+  public static void appendStats(Record toRecord, Record fromRecord) {
+    Preconditions.checkState(toRecord != null, "Record to update cannot be null");
+    Preconditions.checkState(fromRecord != null, "Record to update from cannot be null");
+
+    toRecord.set(
+        Column.SPEC_ID.ordinal(),
+        Math.max(
+            (int) toRecord.get(Column.SPEC_ID.ordinal()),
+            (int) fromRecord.get(Column.SPEC_ID.ordinal())));
+    checkAndIncrementLong(toRecord, fromRecord, Column.DATA_RECORD_COUNT);
+    checkAndIncrementInt(toRecord, fromRecord, Column.DATA_FILE_COUNT);
+    checkAndIncrementLong(toRecord, fromRecord, Column.TOTAL_DATA_FILE_SIZE_IN_BYTES);
+    checkAndIncrementLong(toRecord, fromRecord, Column.POSITION_DELETE_RECORD_COUNT);
+    checkAndIncrementInt(toRecord, fromRecord, Column.POSITION_DELETE_FILE_COUNT);
+    checkAndIncrementLong(toRecord, fromRecord, Column.EQUALITY_DELETE_RECORD_COUNT);
+    checkAndIncrementInt(toRecord, fromRecord, Column.EQUALITY_DELETE_FILE_COUNT);
+    checkAndIncrementLong(toRecord, fromRecord, Column.TOTAL_RECORD_COUNT);
+    if (fromRecord.get(Column.LAST_UPDATED_AT.ordinal()) != null) {
+      if (toRecord.get(Column.LAST_UPDATED_AT.ordinal()) == null
+          || ((long) toRecord.get(Column.LAST_UPDATED_AT.ordinal())
+              < (long) fromRecord.get(Column.LAST_UPDATED_AT.ordinal()))) {
+        toRecord.set(
+            Column.LAST_UPDATED_AT.ordinal(), fromRecord.get(Column.LAST_UPDATED_AT.ordinal()));
+        toRecord.set(
+            Column.LAST_UPDATED_SNAPSHOT_ID.ordinal(),
+            fromRecord.get(Column.LAST_UPDATED_SNAPSHOT_ID.ordinal()));
+      }
+    }
+  }
+
+  private static Record fromManifestEntry(
+      ManifestEntry<?> entry, Table table, Schema recordSchema) {
+    GenericRecord record = GenericRecord.create(recordSchema);
+    Types.StructType partitionType =
+        recordSchema.findField(Column.PARTITION.name()).type().asStructType();
+    Record partitionData = coercedPartitionData(entry.file(), table.specs(), partitionType);
```

Comment:
Does this work with an unpartitioned  spec? or are we guarding against it elsewhere?

Reply:
Currently `recordSchema` was computed by guarded by `PartitionStatsUtil.schema()` and passed to this method. Hence, it was guarded. 

But as it is a public API, better to add the check here also. I will update it.